subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/execution/Lexer.kt

```diff
@@ -17,118 +17,151 @@
+                    return true
+                }
             }
-        } else null
+        }
+        return false
     }
 
+    KotlinLexer().apply {
+        start(script)
+        while (tokenType != null) {
 
-private
-fun KotlinLexer.skipWhiteSpaceAndComments() {
-    while (tokenType in WHITE_SPACE_OR_COMMENT_BIT_SET) {
-        advance()
-    }
-}
+            when (tokenType) {
 
+                WHITE_SPACE -> {
+                    // ignore
+                }
 
-private
-fun KotlinLexer.findTopLevelIdentifier(identifier: String): IntRange? {
-    var depth: Int = 0
-    while (tokenType != null) {
-        when (tokenType) {
-            IDENTIFIER ->
-                if (depth == 0 && tokenText == identifier) {
-                    return tokenStart..(tokenEnd - 1)
+                in COMMENTS -> {
+                    comments.add(
+                        tokenStart..(tokenEnd - 1)
+                    )
                 }
-            LBRACE -> depth += 1
-            RBRACE -> depth -= 1
+
+                else -> {
+
+                    when (state) {
+
+                        State.SearchingTopLevelBlock -> {
+
+                            when (tokenType) {
+                                IDENTIFIER -> matchTopLevelIdentifier()
+                                LBRACE -> depth += 1
+                                RBRACE -> depth -= 1
+                            }
+                        }
+
+                        State.SearchingBlockStart -> {
+
+                            when (tokenType) {
+                                IDENTIFIER -> if (!matchTopLevelIdentifier()) reset()
+                                LBRACE -> {
+                                    depth += 1
+                                    state = State.SearchingBlockEnd
+                                    blockStart = tokenStart
+                                }
+                                else -> reset()
+                            }
+                        }
+
+                        State.SearchingBlockEnd -> {
+
+                            when (tokenType) {
+                                LBRACE -> depth += 1
+                                RBRACE -> {
+                                    depth -= 1
+                                    if (depth == 0) {
+                                        tokens.add(
+                                            topLevelBlock(
+                                                inTopLevelBlock!!,
+                                                blockIdentifier!!,
+                                                blockStart!!..tokenStart
+                                            )
+                                        )
+                                        reset()
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
Consider adding some documentation explaining what you are looking for here.

Reply:
Explaining comment is already present: https://github.com/gradle/kotlin-dsl/pull/887/files#diff-ad9df6da18fa73c22b63aa9b5a7c0921R39

Reply:
You are totally correct, my mistake.